src/capmap.cc

```diff
@@ -0,0 +1,219 @@
+// SPDX-FileCopyrightText: Copyright 2023 Arm Limited and/or its affiliates <open-source-office@arm.com>
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "include/capmap.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+
+// BSD headers for kinfo_getvmmap.
+#include <libutil.h>
+#include <sys/types.h>
+#include <sys/user.h>
+#include <unistd.h>
+
+#if __has_feature(capabilities)
+#include <cheriintrin.h>
+#else
+#error "capmap.cc requires capabilities"
+#endif
+
+namespace capmap {
+
+#ifdef __CHERI_PURE_CAPABILITY__
+#define R8 "c8"
+#else
+#define R8 "x8"
+#endif
+
+__attribute((naked)) Roots get_roots() {
+  // AAPCS64(-cap): The caller allocates space for the large returned struct,
+  // and passes a pointer in x8/c8.
+  // clang-format off
+  asm("1:\n"
+      "stp c0, c1, [" R8 "]\n"
+      "stp c2, c3, [" R8 ", #32]\n"
+      "stp c4, c5, [" R8 ", #64]\n"
+      "stp c6, c7, [" R8 ", #96]\n"
+      "stp c8, c9, [" R8 ", #128]\n"
+      "stp c10, c11, [" R8 ", #160]\n"
+      "stp c12, c13, [" R8 ", #192]\n"
+      "stp c14, c15, [" R8 ", #224]\n"
+      "stp c16, c17, [" R8 ", #256]\n"
+      "stp c18, c19, [" R8 ", #288]\n"
+      "stp c20, c21, [" R8 ", #320]\n"
+      "stp c22, c23, [" R8 ", #352]\n"
+      "stp c24, c25, [" R8 ", #384]\n"
+      "stp c26, c27, [" R8 ", #416]\n"
+      "stp c28, c29, [" R8 ", #448]\n"
+      "str c30, [" R8 ", #480]\n"
+
+      "mov c0, csp\n"
+      "mrs c1, ddc\n"
+      "stp c0, c1, [" R8 ", #496]\n"
+
+#ifdef __CHERI_PURE_CAPABILITY__
+      "adr c0, 1b\n"  // PCC on entry
+#else
+      "adr x0, 1b\n"  // PCC on entry
```

Comment:
Could these two be folded? If there's some `__CHERI_HYBRID_CAPABILITY__` maybe.

Reply:
Folded how? There's no `__CHERI_HYBRID_CAPABILITY__` but it'd bascially be `__CHERI__ && !__CHERI_PURE_CAPABILITY__`, i.e. basically this `#else` clause.

Strictly, this is a C64/A64 difference, and I'm conflating purecap/hybrid with C64/A64, but there's no macro for those, as far as I know.